pgex/examples/maze_generation/generator.py

```diff
@@ -0,0 +1,153 @@
+            [
+                Tile(tilesize, pygame.Vector2(c, r) * tilesize, "black")
+                for c in range(self.width)
+            ]
+            for r in range(self.height)
+        ]
+
+        self.unvisited = [
+            pygame.Vector2(c, r) for r in range(self.height) for c in range(self.width)
+        ]
+        self.definite_walls = []
+        # 0 means wall, 1 means floor
+        self.walls = []
+        self.offsets = [
+            pygame.Vector2(1, 0),
+            pygame.Vector2(-1, 0),
+            pygame.Vector2(0, 1),
+            pygame.Vector2(0, -1),
+        ]
+
+        self.start()
+
+    def start(self) -> None:
+        start_point = (
+            random.randint(0, self.height - 1),
+            random.randint(0, self.width - 1),
+        )
+        self.map[start_point[0]][start_point[1]] = 1
+        self.unvisited.remove(pygame.Vector2(start_point[1], start_point[0]))
+
+        self.add_walls(pygame.Vector2(start_point[1], start_point[0]))
+        while len(self.walls):
+            self.iterate()
+        self.update_tiles()
+
+    def add_walls(self, point: pygame.Vector2) -> None:
+        candidates = [point + offset for offset in self.offsets]
+        for candidate in candidates:
+            if candidate.x < 0 or candidate.y < 0:
+                continue
+
+            elif candidate.x >= self.width:
+                continue
+
+            elif candidate.y >= self.height:
+                continue
+
+            elif candidate not in self.walls and candidate in self.unvisited:
+                self.walls.append(candidate)
+
+    def iterate(self) -> None:
+        point = random.choice(self.walls)
+
+        self.walls.remove(point)
+        visited_neighbors = self.get_visited_neighbors(point)
+        if visited_neighbors == 1 and not self.is_last_wall(point):
+            self.map[int(point.y)][int(point.x)] = 1
+            self.unvisited.remove(point)
+
+            self.add_walls(point)
+        else:
+            self.definite_walls.append(point)
+
+        return point
+
+    def is_last_wall(self, point: pygame.Vector2) -> bool:
+        all_walls = self.unvisited
+
+        neighbors = [point + offset for offset in self.offsets]
+
+        for neighbor in neighbors:
+            count = 0
+            adjacents = [neighbor + offset for offset in self.offsets]
+
+            for adjacent in adjacents:
+                if adjacent in all_walls:
+                    count += 1
+
+            if count <= 1:
+                return True
+
+        return False
+
+    def get_visited_neighbors(self, point: pygame.Vector2):
+        visited = 0
+        candidates = [point + offset for offset in self.offsets]
+        for candidate in candidates:
+            if candidate.x < 0 or candidate.y < 0:
+                continue
+
+            elif candidate.x >= self.width:
+                continue
+
+            elif candidate.y >= self.height:
+                continue
```

Comment:
could shorten to
```py
if not 0 <= candidate.x < self.width and not 0 <= candidate.y < self.height:
    continue
```
also that `continue` is a bit redundant since `if/elif` so could
also do (to replace + 2 lines below)
```py
if 0 <= candidate.x < self.width and 0 <= candidate.y < self.height and candidate not in self.unvisited:
    visited += 1
```